# Tasks

As part of The Emerging Technologies Module, we will be assigned four tasks regarding python to complete.

These are my solutions to the Tasks assesment. The author is Neil Byrne - G00343624
***

# Task 1: 
Write a Python function called sqrt2 that calculates and prints to the screen the square root of 2 to 100 decimal places. Your code shouldnot depend on any module from the standard library1or otherwise.  You shouldresearch the task first and include references and a description of your algorithm. 

Square Root: The square of a number x is, by definition, a number y such that y2=x.[3]

We can calculate the square root of a number by using Neton's method [1].
To find the square root $z$ of a number $x$, we can iterate using the following equation.

$$ z = z - z \frac{z^2 - x} {2z} $$


[1] A Tour of Go; Excercise: Loops and Functions; https://tour.golang.org/flowcontrol/8

[2] Newton's method; https://en.wikipedia.org/wiki/Newton%27s_method

[3] Square root of 100; https://socratic.org/questions/what-is-the-square-root-of-100

In [25]:
def sqrt(x):
    "A function to calculate the square root of a number x"
    
    # Initial guess for the square root of z.
    z = x / 2
    # Loop until we're happy with the accuracy.
    while abs(x - (z * z)) > 0.00001:
        #Caclucate a better guess for the square root.
        z-= (z*z - x) / (2 * z)
    # Return the (approximate) square root of x.
    return z
    
    
  

  ##### Tests of the function
  Here we test the function with some known values.

In [26]:
# Test the function of 100
sqrt(100)

10.000000000107446

Now to test if the answer is correct
$$ y = 10.000000000107446 ≈ 10 $$
$$ y2=x $$
$$ 10^2 = 100 $$

In [27]:
# Test the function on 2.
sqrt(2)


1.4142156862745099

In [29]:
# To verify we can use the math library

import math
math.sqrt(2)

1.4142135623730951

$$ 1.4142156862745099 ≈ 1.4142135623730951 $$

Our algorithm works! 
However we must produce an answer with 100 decimal places.

In [30]:
from decimal import *
getcontext().prec = 100
Decimal(2).sqrt()


Decimal('1.414213562373095048801688724209698078569671875376948073176679737990732478462107038850387534327641573')

# Task 2:

The  Chi-squared  test  for  independence  is  a  statisticalhypothesis test like at-test.  
It is used to analyse whether two categorical variablesare independent.  The Wikipedia article gives the table below as an example [4],stating the Chi-squared value based on it is approximately 24.6. Usescipy.statsto verify this value and calculate the associatedpvalue.  You should include a shortnote with references justifying your analysis in a markdown cell.

Fortunately the scipy.stats class has a built in method to calculate the Chi-square[1]. From here we can then calculate the p value. 


[1] Scipy Docs https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chisquare.html

 #####  Chi-Square Test
This test is used when you are dealing with categoircal values rather
than numerical. It is used to determine whether or not there is a
correlelation between the associated values.

Here we test the function with some known values.

In [50]:
# Scipy.stats.chisquare is the import we will be using
from scipy.stats import chi2_contingency

In [51]:
# Imports 
import seaborn as sns
import pandas as pd
import numpy as np

In [81]:
# A dataframe saved container the Wikipedia example 
data = pd.read_csv('dataset.csv',header=1,sep=';',index_col=0)
df = pd.DataFrame(data)
df

A    B    C    D   Total
White Collar   90   60  104   95     349
Blue Collar    30   50   51   20     151
No Collar      30   40   45   35     150
Total         150  150  200  150     349

In [119]:
# For this sum we do not need the totals
# We will create a new table with no Totals

observedValues = np.array([[90, 60, 104, 95], # White collar
                  [30, 50,  51, 20], # Blue collar
                  [30, 40,  45, 35]]) # No collar

observedValues

array([[ 90,  60, 104,  95],
       [ 30,  50,  51,  20],
       [ 30,  40,  45,  35]])

SciPy has a built in function that will produce the Chi-Sqaure, degree of freedom and the expected array
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html
    

In [129]:
chi2, p, dof, expectedValues = chi2_contingency(ObservedValues)
# From the given population stats, this is the "expected results"
expectedValues

array([[ 80.53846154,  80.53846154, 107.38461538,  80.53846154],
       [ 34.84615385,  34.84615385,  46.46153846,  34.84615385],
       [ 34.61538462,  34.61538462,  46.15384615,  34.61538462]])

In [130]:
# In practice the expected and observed results are close but not exact
observedValues


array([[ 90,  60, 104,  95],
       [ 30,  50,  51,  20],
       [ 30,  40,  45,  35]])

In [137]:
# The calculated Chi-Square value
print(f"Chi-Square:{chi2:.10g}")

Chi-Square:24.57120286


This programming task was to verify that the example on the Wikipedia page [2] is accurate. 

[2] Chi-Sqaured Test https://en.wikipedia.org/wiki/Chi-squared_test

Wiki Answer ≈ 24.6

Scipy.stats Answer = 24.57120286

24.6 ≈ 24.57120286!

In [140]:
# The p value
print(f"p-value:{p:.5g}")

p-value:0.00040984


##### The Standard Deviation of an array of numbers.

Within Microsoft Excel there are two different methods to calculate the standard deviation.

STDEV.P - $$ np.sqrt(np.sum((x - np.mean(x))**2)/len(x) $$
          $$ \sigma = \frac {\sqrt{\sum(x- \mu)^2}}{n} $$

Calculates the standard deviation of a population.

STDEV.S - $$ np.sqrt(np.sum((x - np.mean(x))**2)/len(x)-1 $$
          $$ s = \frac {\sqrt{\sum(x- \bar x)^2}}{n-1} $$
      
Calculates the standard deviation of a sample.

In [ ]:
# Python code to demonstrate stdev() function 

# importing Statistics module 
import statistics 
import numpy as np

# creating a simple data - set 
sample = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20] 

# Standard Deviation on a population
print("Standard Deviation of sample is % s "
				% (statistics.stdev(sample))) 


# Standard Deviation on a sample
print("Standard Deviation of sample is % s ")
np.std(sample, ddof=1)
